utmpx: add some accounting-related POSIX functions

lucab opened the pull request:
This adds initial support for reading `utmpx` entries, as defined by
POSIX.

Ref: http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/utmpx.h.html
Closes: #1111

lucab commented:
@asomers this should be ready for an initial review pass.

lucab commented on src/utmpx.rs:
```diff
@@ -0,0 +1,275 @@
+/// Valid `UtmpEntry` entry types.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, Ord, PartialOrd)]
+#[repr(i16)]
+pub enum EntryType {
```
Thanks, I wasn't aware of it. It doesn't seem to cover variant-with-explicit-discriminant syntax though, or am I just not seeing it? Assuming it does not, mind if I leave the manual impl in?

asomers replied:
No, `libc_enum` should work fine.  It doesn't implement `TryFrom` _yet_, but I'll take care of that soon.

asomers replied:
Nevermind, I won't be implementing TryFrom within libc_enum afterall.  It's too hard to do it in a regular macro and I don't want to go through the boilerplate needed to add proc macros right now.

lucab commented on src/utmpx.rs:
```diff
@@ -0,0 +1,275 @@
+
+    #[cfg(not(target_env = "gnu"))]
+    /// Set default path (unsupported GNU extension).
+    fn set_default_path() -> Result<CString> {
```
This "sets" the underlying db file to use, as much as `setutxdb` does (yes, the underlying API is terrible)
If that's still not fine, what would be your preferred name for this?

lucab commented on src/lib.rs:
```diff
@@ -72,6 +77,12 @@ pub mod sys;
 pub mod unistd;
+#[deny(missing_docs)]
+#[cfg(any(target_env = "gnu",
+          target_os = "freebsd",
```
I'm happy to make this as portable as possible, but I'd rather not block this PR on future `libc` API. Would you mind if we gradually wire-in other platforms later on?

lucab commented on test/test_utmpx.rs:
```diff
@@ -0,0 +1,47 @@
+        let mut handle = stderr.lock();
+        writeln!(
+            handle,
+            "empty utmp database, test results may be unreliable"
```
I'm confused. I think `write!` does not, but `writeln!` should. Am I misunderstanding something?

asomers replied:
No, you're right and I'm wrong.  But why do you need to lock it?  Can't you do `writeln!(&mut std::io::stderr, "...")`?

asomers commented on src/utmpx.rs:
```diff
@@ -238,6 +270,7 @@ mod tests {
+    #[cfg(any(target_env = "gnu", target_os = "dragonfly", target_os = "freebsd"))]
     #[test]
     /// Test a bare get entry.
     fn test_getutxent() {
```
What coverage does this test provide that the integration test doesn't?

lucab replied:
No additional coverage. But this helps sanity checking the C primitives, especially on the CI here where I don't have insights otherwise.

asomers replied:
I think this test should be deleted, then.  It's not Nix's job to write a test suite for the utmpx(3) functions.  The integration test is good enough for Nix's purposes.

lucab replied:
Ack. Should I delete all the unit tests then? Or were you specifically concerned with this one only?

asomers replied:
Yeah, you should probably delete them all.

lucab commented on src/utmpx.rs:
```diff
@@ -181,20 +192,34 @@ impl Utmp {
+    fn set_default_path() -> Result<()> {
+        use std::ffi::CString;
+
+        let path = CString::new(UTMPX_DEFAULT_PATH).map_err(|_| Error::InvalidPath)?;
```
I've left a TODO here, I'll followup on this. It is a string only used internally, if possible I'd rather not block this on future libc API.

lucab replied:
I've added a TODO comment for this, I'm gradually following up on libc. If possible, I'd rather not block this on future libc API.

asomers replied:
No, I really don't want to define any magic strings in Nix.  That string is part of the `utmpx` internal implementation, not its public API.  It will only cause problems to define it here.

lucab commented on src/utmpx.rs:
```diff
@@ -159,14 +170,14 @@ impl<'a> Iterator for UtmpIter<'a> {
 #[derive(Clone, PartialEq, Eq, Debug)]
 pub struct Utmp {
-    path: CString,
+    _private: (),
```
I'm using this just to prevent consumer from directly building this type. The `open` contructor has proper additional logic for rewinding on open. Any alternative suggestion?

lucab replied:
I'm using this just to prevent consumer from directly building this type. The `open` contructor has proper additional logic for rewinding on open. Any alternative suggestion?

asomers replied:
That makes sense.  Could you please add a comment to that effect?

lucab commented on src/utmpx.rs:
```diff
@@ -0,0 +1,275 @@
+    /// Set default path (GNU extension).
+    fn set_default_path() -> Result<CString> {
+        let path = CString::new(UTMPX_FILENAME).map_err(|_| Error::InvalidPath)?;
+        let res = unsafe { libc::utmpxname(path.as_ptr()) };
```
Unfortunately libc only exposes `utmpxname` right now (not the bare one). `utmpxname(3)` manpage exists though, mentioning it is a glibc extension.

asomers replied:
Yeah, but the man page doesn't say anything about what it does, just that's it's a glibc extension.  That's why I suggested to use the documented `utmpname` instead.

lucab replied:
Ack. The implementation is a [bare wrapper](https://sourceware.org/git/?p=glibc.git;a=blob;f=login/utmpxname.c;h=0631ca56cda81c1590ae7a4bcb8261f2350b8136;hb=HEAD). Do you prefer me to open a PR to `libc` to add the other function, or to the glibc manpage to explicitly mention they have the exact same semantics?

asomers replied:
Better to open a PR to add `utmpname`.

lucab replied:
Followed up with https://github.com/rust-lang/libc/pull/1494.

lucab commented:
@asomers thanks for the very quick reviews! I think I've addressed all your comments so far.
I've several PRs in flight on `libc` side, and relevant links in TODO items.
May I squash all the fixups?

asomers commented:
Why the rush?  Let's just wait until all of the libc PRs get merged.

lucab commented:
I've found some time to come back and rebase this PR, using the constants that landed in `libc` in the meanwhile. @asomers PTAL.

asomers commented on src/utmpx.rs:
```diff
@@ -0,0 +1,234 @@
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct UtmpEntry {
+    ut_host: String,
+    ut_id: Vec<u8>,
```
Since it has a fixed (platform-dependent) size, you can store `ut_id` as an array instead of a vec.

lucab replied:
To the best of my knowledge, the size of this field is not a public API and there is no stable `libc` constant defining it.
But I fear I may be missing some bits of your suggestion. If you have clues in other directions, and actual working declaration you want me to use here, I'll be happy to follow.

asomers commented on src/utmpx.rs:
```diff
@@ -0,0 +1,234 @@
+    ///
+    /// # Safety
+    ///
+    /// This operations is unsafe because it mutates global libc state. In order to
```
s/operations/operation/

asomers commented on src/utmpx.rs:
```diff
@@ -0,0 +1,234 @@
+    /// This operations is unsafe because it mutates global libc state. In order to
+    /// safely invoke this, the caller must ensure that nothing else in the process
+    /// is accessing the `utmp` database at the same time.
+    pub unsafe fn rewind(&mut self) {
```
I think this method can be safe, right?  Because it requires a `Utmp` instance, it can't be invoked without the caller already doing something `unsafe`.  And the `&mut` reference will prevent it from being invoked while there is an active `UtmpIter`.  The only problem I can foresee would be if the caller created two `Utmp` instances, but that would violate the rules for `Utmp::open`.

asomers commented on test/test_utmpx.rs:
```diff
@@ -0,0 +1,41 @@
+/// Test iterating through default DB.
+#[test]
+#[cfg(all(
+    any(target_arch = "x86", target_arch = "x86_64"),
```
You should add a comment explaining why this test is only enabled for certain architectures.

lucab commented:
@asomers I'm sorry for the noise, but I don't have a good way to rebase-on-master and play on my branch with your funky CI setup without triggering a notification. For the time being I'm still reworking and experimenting with this. I'll ping you when it's ready for another review round.

lucab commented on src/sys/time.rs:
```diff
@@ -249,6 +249,18 @@ impl fmt::Display for TimeSpec {
 pub struct TimeVal(timeval);
 
+impl TimeVal {
+    #[allow(dead_code)]
```
The caller is behind a platform-conditional, some CI platforms fail without this.
The logic here is not platform-dependent though, thus allowing dead-code for this specific scope.

lucab commented on src/utmpx.rs:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    /// Copy out buffer content, as a string, up to the first NUL-terminator.
+    fn charbuf_to_string(input: &[libc::c_char]) -> Result<String> {
```
It's a different semantic though: the logic here adds a safety terminator, the stdlib logic forbids internal NULs (see `NulError`).

asomers replied:
According to FreeBSD's man page, the `ut_user`, `ut_line`, and `ut_host` fields are guaranteed to be NUL-terminated.  So you don't need to add your own termination, now that you're no longer treating ut_id as a string.

lucab commented on test/test_utmpx.rs:
```diff
@@ -0,0 +1,45 @@
+        }
+    }
+
+    // Cross-jobs in Travis seems to have an empty DB.
```
I've tried a few combinations of target-conditions (the one quoted here was actually failing in CI), this seems to affect all Travis test-runs (including x86_64), I suspect it may be related to a containerized/virtualized step somewhere.
In additional tests, the annotation suggested in https://github.com/nix-rust/nix/blob/master/CONTRIBUTING.md#disabling-a-test-in-the-ci-environment did not really work to ignore it here.

I've temporarily put an `ignore` annotation here with a FIXME note. The tests itself should be fine (it works locally), but the CI needs some tweaks on Travis side to cope with it.

lucab commented on src/utmpx.rs:
```diff
@@ -0,0 +1,209 @@
+        let entry = unsafe { *ptr };
+        let res = Self {
+            ut_user: Self::charbuf_to_string(&entry.ut_user)?,
+            ut_id: Self::charbuf_to_string(&entry.ut_id)?,
```
My understanding is that the fields are always present, but the meaning of their content may vary. I'm using https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/utmpx.h.html as my base reference here.

Would you mind linking/referencing the docpage you are looking at?

For reference, at least on Linux those fields have meaningful content on each entry  regardless of the type, for example:
```
UtmpEntry { ut_host: "4.19.0-8-amd64", ut_id: [126, 126, 0, 0], ut_line: "~", ut_pid: Pid(0), ut_tv: TimeVal(timeval { tv_sec: 1590841230, tv_usec: 500327 }), ut_type: BOOT_TIME, ut_user: "reboot" }
UtmpEntry { ut_host: "4.19.0-8-amd64", ut_id: [126, 126, 0, 0], ut_line: "~", ut_pid: Pid(53), ut_tv: TimeVal(timeval { tv_sec: 1590841244, tv_usec: 231680 }), ut_type: RUN_LVL, ut_user: "runlevel" }
```

asomers replied:
Here's the man page that describes those fields as not relevant for all event types.
https://www.freebsd.org/cgi/man.cgi?query=getutxent&apropos=0&sektion=0&manpath=FreeBSD+12.1-RELEASE+and+Ports&arch=default&format=html
And here is the output of `last` on FreeBSD.  Notice that the line and host columns are blank for the boot time entry:
```
somers     pts/0    threonine.lauralan.noi Sat May 30 12:25   still logged in
boot time                                  Wed May 27 13:12
```

lucab commented:
@asomers I think I've applied most of your comments and replied to the ones that didn't directly apply. PTAL and thanks for the ongoing feedback (I've tried to isolate the changes in a fixup commit while still rebasing on master).

asomers commented on src/utmpx.rs:
```diff
@@ -0,0 +1,229 @@
+
+        // The lifetime of this whole buffer is very shady and the overall
+        // content small, so we copy the data out and start owning it.
+        let entry = unsafe { *ptr };
```
The lifetime can't possibly be shorter than the remainder of this function, or else there will be very big problems.
